source/tests/SGM/SGMMonochromatorInfoTestView.cpp

```diff
@@ -0,0 +1,428 @@
+    makeConnections();
+}
+
+void SGMMonochromatorInfoTestView::onMonoGratingAngleChanged(double gratingAngleSteps)
+{
+    gratingAngleSpinBox_->setValue(gratingAngleSteps);
+}
+
+void SGMMonochromatorInfoTestView::onMonoGratingTranslationChanged(SGMGratingSupport::GratingTranslation gratingTranslation)
+{
+    int gratingInteger = int(gratingTranslation);
+
+    gratingTranslationComboBox_->setCurrentIndex(gratingInteger);
+}
+
+void SGMMonochromatorInfoTestView::onMonoUndulatorTrackingChanged(bool isTracking)
+{
+    undulatorTrackingCheckBox_->setChecked(isTracking);
+}
+
+void SGMMonochromatorInfoTestView::onMonoUndulatorPositionChanged(double undulatorPosition)
+{
+    undulatorPositionSpinBox_->setValue(undulatorPosition);
+}
+
+void SGMMonochromatorInfoTestView::onMonoUndulatorOffsetChanged(double undulatorOffset)
+{
+    undulatorDetuneOffsetSpinBox_->setValue(undulatorOffset);
+}
+
+void SGMMonochromatorInfoTestView::onMonoUndulatorHarmonicChanged(SGMUndulatorSupport::UndulatorHarmonic undulatorHarmonic)
+{
+    switch(undulatorHarmonic) {
+    case SGMUndulatorSupport::FirstHarmonic:
+        undulatorHarmonicComboBox_->setCurrentIndex(0);
+        break;
+    case SGMUndulatorSupport::ThirdHarmonic:
+        undulatorHarmonicComboBox_->setCurrentIndex(1);
+        break;
+    case SGMUndulatorSupport::UnknownUndulatorHarmonic:
+        break;
+    }
+}
+
+void SGMMonochromatorInfoTestView::onMonoExitSlitTrackingChanged(bool isTracking)
+{
+    exitSlitTrackingCheckBox_->setChecked(isTracking);
+}
+
+void SGMMonochromatorInfoTestView::onMonoExitSlitPositionChanged(double exitSlitPosition)
+{
+    exitSlitPositionSpinBox_->setValue(exitSlitPosition);
+}
+
+void SGMMonochromatorInfoTestView::onMonoEnergyChanged(double energy)
+{
+    energySpinBox_->setValue(energy);
+}
+
+void SGMMonochromatorInfoTestView::onMonoWarningCountChanged(int warningCount)
+{
+    if(warningCount == 0) {
+        warningsTextEdit_->setVisible(false);
+    } else {
+
+        warningsTextEdit_->setVisible(true);
+        QString warningString;
+
+        foreach(QString message, sgmMonochromatorInfo_->warningMessages()) {
+
+            warningString.append(QString("\n%1").arg(message));
+        }
+
+        warningsTextEdit_->setText(warningString);
+    }
+
+}
+
+void SGMMonochromatorInfoTestView::onMonoErrorCountChanged(int errorCount)
+{
+    if(errorCount == 0) {
+        errorsTextEdit_->setVisible(false);
+    } else {
+
+        errorsTextEdit_->setVisible(true);
+        QString errorString;
+
+        foreach(QString message, sgmMonochromatorInfo_->errorMessages()) {
+
+            errorString.append(QString("\n%1").arg(message));
+        }
+
+        errorsTextEdit_->setText(errorString);
+    }
+}
+
+void SGMMonochromatorInfoTestView::onGratingAngleSpinBoxChanged()
+{
+    double value = gratingAngleSpinBox_->value();
+    if(qAbs(value - sgmMonochromatorInfo_->gratingAngle()) > 0.001 ||
+            sgmMonochromatorInfo_->hasErrors()) {
```

Comment:
Is this logically correct?  You want to set a new value to the info if the value is different (seems fine) _OR_ there are errors?